src/js/asymmetric.js

```diff
@@ -0,0 +1,112 @@
+/*
+
+    An "asymmetric" dataSource designed to read from a CouchDB view, list, show, etc. and to write to CouchDB's
+    document API. See the documentation for details:
+
+    https://github.com/the-t-in-rtf/gpii-couch-cushion/blob/GPII-1253/docs/asymmetric.md
+
+ */
+"use strict";
+var fluid = fluid || require("infusion");
+
+require("kettle");
+
+require("./lib/transformEach");
+
+fluid.defaults("gpii.couchdb.cushion.asymmetricDataSource", {
+    gradeNames: ["fluid.component"],
+    events: {
+        onRead:       null,
+        onReadError:  null,
+        onWrite:      null,
+        onWriteError: null
+    },
+    readTermMap:  {},
+    writeTermMap: {
+        _id: "%_id"
+    },
+    components: {
+        getter: {
+            type: "kettle.dataSource.URL",
+            options: {
+                gradeNames: "kettle.dataSource.CouchDB",
+                writable: false,
+                model: {
+                    getModel: true
+                },
+                rules: {
+                    // The default rules pull out the inner values from a CouchDB list.  See the docs for examples.
+                    readPayload: {
+                        "rows": {
+                            transform: {
+                                type: "fluid.transforms.transformEach",
+                                inputPath: "rows",
+                                rules: {
+                                    "": "value"
+                                }
+                            }
+                        }
+                    }
+                },
+                url:         "{gpii.couchdb.cushion.asymmetricDataSource}.options.readUrl",
+                termMap:     "{gpii.couchdb.cushion.asymmetricDataSource}.options.readTermMap",
+                listeners: {
+                    "onError.notifyParent": {
+                        func: "{gpii.couchdb.cushion.asymmetricDataSource}.events.onReadError.fire"
+                    },
+                    "onRead.notifyParent": {
+                        func: "{gpii.couchdb.cushion.asymmetricDataSource}.events.onRead.fire"
+                    }
+                }
+            }
+        },
+        setter: {
+            type: "kettle.dataSource.URL",
+            options: {
+                gradeNames: "kettle.dataSource.CouchDB",
+                writable: true,
+                model: {
+                    setModel: true
+                },
+                url:         "{gpii.couchdb.cushion.asymmetricDataSource}.options.writeUrl",
+                termMap:     "{gpii.couchdb.cushion.asymmetricDataSource}.options.writeTermMap",
+                listeners: {
+                    "onError.notifyParent": {
+                        func: "{gpii.couchdb.cushion.asymmetricDataSource}.events.onWriteError.fire"
+                    },
+                    "onWrite.notifyParent": {
```

Comment:
https://github.com/amb26/kettle/tree/KETTLE-32#transforming-promise-chains

On 31/03/2016 12:43, Tony Atkins wrote:
> In src/js/asymmetric.js <https://github.com/GPII/gpii-couch-cushion/pull/1#discussion_r58039831>:
>
>> +        },
>> +        setter: {
>> +            type: "kettle.dataSource.URL",
>> +            options: {
>> +                gradeNames: "kettle.dataSource.CouchDB",
>> +                writable: true,
>> +                model: {
>> +                    setModel: true
>> +                },
>> +                url:         "{gpii.couchdb.cushion.asymmetricDataSource}.options.writeUrl",
>> +                termMap:     "{gpii.couchdb.cushion.asymmetricDataSource}.options.writeTermMap",
>> +                listeners: {
>> +                    "onError.notifyParent": {
>> +                        func: "{gpii.couchdb.cushion.asymmetricDataSource}.events.onWriteError.fire"
>> +                    },
>> +                    "onWrite.notifyParent": {
>
> Thanks. Is this documented anywhere, even in draft form?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GPII/gpii-couch-cushion/pull/1/files/701fbbd4aade9bb5c118c3fb3362d88cd30eda15#r58039831>
>